src/EditorFeatures/Core/Implementation/IntelliSense/AsyncCompletion/CommitManager.cs

```diff
@@ -0,0 +1,196 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Linq;
+using System.Threading;
+using Microsoft.CodeAnalysis.Completion;
+using Microsoft.CodeAnalysis.Editor.Shared.Extensions;
+using Microsoft.CodeAnalysis.Shared.Extensions;
+using Microsoft.CodeAnalysis.Text;
+using Microsoft.CodeAnalysis.Text.Shared.Extensions;
+using Microsoft.VisualStudio.Language.Intellisense.AsyncCompletion;
+using Microsoft.VisualStudio.Text;
+using Microsoft.VisualStudio.Text.Editor;
+using Roslyn.Utilities;
+using AsyncCompletionData = Microsoft.VisualStudio.Language.Intellisense.AsyncCompletion.Data;
+using RoslynCompletionItem = Microsoft.CodeAnalysis.Completion.CompletionItem;
+using VSCompletionItem = Microsoft.VisualStudio.Language.Intellisense.AsyncCompletion.Data.CompletionItem;
+
+namespace Microsoft.CodeAnalysis.Editor.Implementation.IntelliSense.Completion.AsyncCompletion
+{
+    internal sealed class CommitManager : IAsyncCompletionCommitManager
+    {
+        private static readonly IEnumerable<char> s_commitChars = ImmutableArray.Create(
+            ' ', '{', '}', '[', ']', '(', ')', '.', ',', ':',
+            ';', '+', '-', '*', '/', '%', '&', '|', '^', '!',
+            '~', '=', '<', '>', '?', '@', '#', '\'', '\"', '\\');
+
+        private static readonly AsyncCompletionData.CommitResult CommitResultUnhandled =
+            new AsyncCompletionData.CommitResult(isHandled: false, AsyncCompletionData.CommitBehavior.None);
+
+        public IEnumerable<char> PotentialCommitCharacters => s_commitChars;
```

Comment:
what are PotentialCommitCharacters for?

Reply:
// Summary:
        //     Returns characters that may commit completion.
        //     When completion is active and a text edit matches one of these characters, Microsoft.VisualStudio.Language.Intellisense.AsyncCompletion.IAsyncCompletionCommitManager.ShouldCommitCompletion(Microsoft.VisualStudio.Language.Intellisense.AsyncCompletion.IAsyncCompletionSession,Microsoft.VisualStudio.Text.SnapshotPoint,System.Char,System.Threading.CancellationToken)
        //     is called to verify that the character is indeed a commit character at a given
        //     location.
        //     Called on UI thread.

---
In reply to: [222519390](https://github.com/dotnet/roslyn/pull/29016#discussion_r222519390) [](ancestors = 222519390)

Reply:
but what happens when a completion item wants to add a commit character that isn't in that list?  

The idea with completion is that there's no way to know this information independent of a completion item.  Because it's up to the completion item to define the characters that commit.  You can't know them apriori...

Reply:
If we'll not be able to maintain the aggregate list, we will have to ask the editor to ignore PotentialCommitCharacters

---
In reply to: [222525171](https://github.com/dotnet/roslyn/pull/29016#discussion_r222525171) [](ancestors = 222525171)

Reply:
Changed: now we get those characters from CompletionService. Extensions will be able to override us.

---
In reply to: [223928094](https://github.com/dotnet/roslyn/pull/29016#discussion_r223928094) [](ancestors = 223928094)